### PR TITLE
LOK-2221: Add nodeCount grpc method

### PR DESCRIPTION
**Pull request by MusaidAli**
## Jira link(s)
- 1.	https://opennms.atlassian.net/browse/LOK-2221
- 

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: inventory/main/src/main/java/org/opennms/horizon/inventory/repository/NodeRepository.java [hunk 1]
```diff
@@ -46,7 +46,8 @@ public interface NodeRepository extends JpaRepository<Node, Long> {
     Optional<Node> findByIdAndTenantId(long id, String tenantID);
     List<Node> findByNodeLabel(String label);
     List<Node> findByTenantIdAndMonitoredStateEquals(String tenantId, MonitoredState monitoredState);
-
+    @Query("SELECT COUNT(DISTINCT n.id) FROM Node n")
+    long countDistinctNodes();
```
Comment (smunir-onms) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/repository/NodeRepository.java:50]: I think, it should filter by tenantID.

@cgorantla , please correct me if I am wrong.

File: rest-server/src/main/java/org/opennms/horizon/server/service/GrpcNodeService.java [hunk 2]
```diff
@@ -174,7 +174,10 @@ public Mono<TopNResponse> downloadTopN(@GraphQLEnvironment ResolutionEnvironment
                 }
             });
     }
-
+    @GraphQLQuery(name= "countNodes" )
+    public Mono<Long> countNodeRows(@GraphQLEnvironment ResolutionEnvironment env) {
```
Comment (cgorantla) [inline rest-server/src/main/java/org/opennms/horizon/server/service/GrpcNodeService.java:178]: There is an existing GraphQL API for this.
Please use that https://github.com/OpenNMS-Cloud/lokahi/pull/1615/files#diff-c3aa9df5bed1ef1085d4017c23e43a65ff52de5be3e873ee8f1794e821eee498L154

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java [hunk 3]
```diff
@@ -415,4 +416,17 @@ private void sendScannerTasksToMinion(Map<Long, List<NodeDTO>> locationNodes, St
             scannerService.sendNodeScannerTask(entry.getValue(), entry.getKey(), tenantId);
         }
     }
+    @Override
+    public void getNodesCount(Empty request,StreamObserver<Int64Value> responseObserver) {
+        try {
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java:421]: Need to add tenant lookup for every service.

String tenantId = tenantLookup.lookupTenantId(Context.current()).orElseThrow();

`nodeCount` would be better name for the service.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/repository/NodeRepository.java (hunk 1 above)
Comment (cgorantla) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/repository/NodeRepository.java:50]: Yes, we need to get num of nodes for the given tenant

**Review by cgorantla — CHANGES_REQUESTED** [review]

File: shared-lib/inventory/src/main/proto/nodeService.proto [hunk 4]
```diff
@@ -117,4 +117,5 @@ service NodeService {
   rpc startNodeScanByIds(NodeIdList) returns(google.protobuf.BoolValue) {};
   rpc getIpInterfaceById(google.protobuf.Int64Value) returns (IpInterfaceDTO) {};
   rpc getIpInterfaceFromQuery(NodeIdQuery) returns (IpInterfaceDTO) {};
+  rpc getNodesCount(google.protobuf.Empty) returns (google.protobuf.Int64Value) {};
```
Comment (cgorantla) [inline shared-lib/inventory/src/main/proto/nodeService.proto:120]: nit: 
```suggestion
  rpc getNodeCount(google.protobuf.Empty) returns (google.protobuf.Int64Value) {};
```

File: inventory/main/src/main/java/org/opennms/horizon/inventory/repository/NodeRepository.java [hunk 5]
```diff
@@ -46,7 +46,8 @@ public interface NodeRepository extends JpaRepository<Node, Long> {
     Optional<Node> findByIdAndTenantId(long id, String tenantID);
     List<Node> findByNodeLabel(String label);
     List<Node> findByTenantIdAndMonitoredStateEquals(String tenantId, MonitoredState monitoredState);
-
+    @Query("SELECT COUNT(DISTINCT n.id) FROM Node n WHERE n.tenantId = :tenantId ")
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/repository/NodeRepository.java:49]: you can omit DISTINCT as  id is primary key here.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java [hunk 6]
```diff
@@ -415,4 +416,29 @@ private void sendScannerTasksToMinion(Map<Long, List<NodeDTO>> locationNodes, St
             scannerService.sendNodeScannerTask(entry.getValue(), entry.getKey(), tenantId);
         }
     }
+    @Override
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java:419]: Please add a step to validate this in CucumberIT, probably [here](https://github.com/OpenNMS-Cloud/lokahi/blob/460d08408d815314ea212f775df0bcffb27c23c1/inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java#L114) you can replace with  this new grpc method.

**Review by cgorantla — COMMENTED** [review]
Looks good. few more suggestions

**Comment by smunir-onms** [comment]
Thank you so much @cgorantla for the kind suggestions, it's very helpful for future commits.

**Review by cgorantla — APPROVED** [review]
LGTM!

File: inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java [hunk 7]
```diff
@@ -174,5 +174,7 @@ public void fetchCountOfNondes() {
         Int64Value nodeCount;
         nodeCount = nodeServiceBlockingStub.getNodeCount(Empty.newBuilder().build());
 
+        Assert.assertEquals(nodeCount, nodeCount);
```
Comment (cgorantla) [inline inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java:177]: check against actual node count

**Review by cgorantla — COMMENTED** [review]
Please merge develop and handle conflicts

**Pull request closed without merging**